impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaNodeConfiguration.java

```diff
@@ -0,0 +1,327 @@
+    private final int nodeId;
+    private final Set<Role> roles;
+
+    @Getter
+    private final KafkaClusterConfig config;
+
+    @Getter
+    private final String kafkaKraftClusterId;
+    private final String brokerConnectAddress;
+    private final EndpointPair externalEndpoint;
+    private final EndpointPair anonEndpoint;
+    private final EndpointPair internalEndpoint;
+    private final EndpointPair controllerEndpoint;
+
+    public int externalPort() {
+        if (!roles.contains(Role.BROKER)) {
+            throw new IllegalStateException("non-broker node doesn't have an external port");
+        }
+        return externalEndpoint.getConnect().getPort();
+    }
+
+    public int anonPort() {
+        if (!roles.contains(Role.BROKER)) {
+            throw new IllegalStateException("non-broker node doesn't have an anonymous port");
+        }
+        return anonEndpoint.getConnect().getPort();
+    }
+
+    public String clientEndpoint() {
+        if (!roles.contains(Role.BROKER)) {
+            throw new IllegalStateException("non-broker node doesn't have an client accessible endpoint");
+        }
+        return externalEndpoint.connectAddress();
+    }
+
+    public KafkaNodeConfiguration(KafkaTopology kafkaTopology, int nodeId, Set<Role> roles, KafkaClusterConfig config, KafkaEndpoints endpoints) {
+        this.kafkaTopology = kafkaTopology;
+        this.nodeId = nodeId;
+        this.roles = roles;
+        this.config = config;
+        this.kafkaKraftClusterId = config.getKafkaKraftClusterId();
+        if (isBroker()) {
+            EndpointPair endpointPair = endpoints.getEndpointPair(Listener.EXTERNAL, nodeId());
+            brokerConnectAddress = endpointPair.connectAddress();
+        }
+        else {
+            brokerConnectAddress = null;
+        }
+        // allocate endpoints immediately
+        if (roles.contains(Role.BROKER)) {
+            externalEndpoint = endpoints.getEndpointPair(Listener.EXTERNAL, nodeId);
+            anonEndpoint = endpoints.getEndpointPair(Listener.ANON, nodeId);
+            internalEndpoint = endpoints.getEndpointPair(Listener.INTERNAL, nodeId);
+        }
+        else {
+            externalEndpoint = null;
+            anonEndpoint = null;
+            internalEndpoint = null;
+        }
+        if (roles.contains(Role.CONTROLLER)) {
+            controllerEndpoint = endpoints.getEndpointPair(Listener.CONTROLLER, nodeId);
+        }
+        else {
+            controllerEndpoint = null;
+        }
+    }
+
+    /**
+     * Get a broker configs for a specific <code>node.id</code>.
+     *
+     * @return broker configuration.
+     */
+    @NotNull
+    private Properties generateConfig() {
+        Properties nodeConfiguration = new Properties();
+        nodeConfiguration.putAll(config.getBrokerConfigs());
+        putConfig(nodeConfiguration, "broker.id", nodeIdString());
+
+        var protocolMap = new TreeMap<String, String>();
```

Comment:
These maps of listener properties were also a frustration I didn't get to, still to be cleaned up in a future PR.